static/modules/entity.js

```diff
@@ -114,18 +128,76 @@ export class Entity {
     this.vel = vel;
     /** @type {Vector} */
     this.acc = acc;
+    this.collisionType = "Circle";
+  }
+
+  /**
+   * Set the collisionShape of the entity. Give no arguments to reset the shape
+   * to be calculated each call of getCollisonShape
+   * @param {CollisionShape | undefined} collisionShape
+   * @returns {void}
+   */
+  setCollisionShape(collisionShape) {
+    this.collisionShape = collisionShape;
+  }
+  /**
+   * Set the collisionShape of the entity when colliding with terrain. Give no
+   * arguments to reset the shape to be the same as collisionShape
+   * @returns {void}
+   */
+  setTerrainCollisionShape(terrainCollisionShape) {
+    this.terrainCollisionShape = terrainCollisionShape;
+  }
+
+  /**
+   * Returns the CollisionShape of the entity.
+   * If the entity has a collisionEntity property, that shape is updated to the
+   * current position and is returned. Otherwise, a new CollisionShape is
+   * calculated.
+   * @returns {CollisionShape}
+   */
+  getCollisionShape() {
+    if (this.collisionShape !== undefined) {
+      this.collisionShape.pos = this.pos;
+      this.collisionShape.vel = this.vel;
+      return this.collisionShape;
+    } else if (this.collisionType == "Box") {
+      return new Box(this.width, this.height, this.pos, this.vel);
+    } else if (this.collisionType == "Circle") {
+      return new Circle(
+        Math.min(this.width, this.height) / 2,
+        this.pos,
+        this.vel
+      );
+    }
+    return new CollisionShape("undefined", this.pos, this.vel);
+  }
+
+  /**
+   * Returns the terrainCollisionShape of the entity.
+   * If the entity has a collisionEntity property, that shape is updated to the
+   * current position and returned. Otherwise, the collisionShape is returned.
+   * @returns {CollisionShape}
+   */
+  getTerrainCollisionShape() {
+    if (this.terrainCollisionShape !== undefined) {
+      this.terrainCollisionShape.pos = this.pos;
+      this.terrainCollisionShape.vel = this.vel;
+      return this.terrainCollisionShape;
+    }
+    return this.getCollisionShape();
   }
 
   onScreen() {
     const { width: screenWidth, height: screenHeight } = getScreenDimensions();
-    const screenEntity = new Entity(
+    const screenBox = new Box(
+      screenWidth,
+      screenHeight,
       new Vector(screenWidth / 2, screenHeight / 2).add(
         getCameraOffset().mult(-1)
       )
     );
-    screenEntity.width = screenWidth;
-    screenEntity.height = screenHeight;
-    return isColliding(this, screenEntity);
+    return !collide(this.getCollisionShape(), screenBox).isZeroVec() || true;
```

Comment:
ummmm... this always returns true??

does this mean you've been testing with every entity being reported as on screen?

Reply:
Oops. Fixed in new commit
Revealed a bug where entities were being falsely reported as not colliding when the two entities had no velocity, which was fixed.